### PR TITLE
(Fixes #500) Support for attachments in Slack notification

### LINKED ISSUE #500: Support for attachments in Slack notifier
Provide support for attachments in Slack messages : https://api.slack.com/docs/message-attachments

This would allow messages to be highlighted in `red-danger`, `green-good`, `yellow-warning` or any other custom color required by the user.

Helpful to filter out FATAL messages in a slack channel flooded with alerts.

### PR DESCRIPTION
**Pull request by ssharma**
* Support for attachments in Slack notifications

**Tested for Attachments**

```scala
  def main(arg: Array[String]): Unit = {
    var args = new Array[String](8)
    args(0) = "--webhook-url"
    args(1) = "my web hook url"
    args(2) = "--useAttachment"
    args(3) = "--title"
    args(4) = "Pipeline Failed"
    args(5) = "--color"
    args(6) = "danger"
    args(7) = "I KNOW IT FAILED!"

    val parser = new OptionParser[Options](s"hyperion-notification-slack-activity") {
      override def showUsageOnError = true

      note("Sends a notification message to a Slack incoming webhook.")
      help("help").text("prints this usage text")
      opt[Unit]("fail-on-error").optional().action((_, c) => c.copy(failOnError = true))
        .text("Causes the activity to fail if any error received from the webhook")
      opt[String]("webhook-url").valueName("WEBHOOK").required().action((x, c) => c.copy(webhookUrl = x))
        .text("Sends the message to the given WEBHOOK url")
      opt[String]("user").valueName("NAME").optional().action((x, c) => c.copy(user = Option(x)))
        .text("Sends the message as the user with NAME")
      opt[String]("emoji").valueName("EMOJI").optional().action((x, c) => c.copy(iconEmoji = Option(x)))
        .text("Use EMOJI for the icon")
      opt[String]("to").valueName("CHANNEL or USERNAME").optional().action((x, c) => c.copy(channel = Option(x)))
        .text("Sends the message to #CHANNEL or @USERNAME")
      opt[Unit]("useAttachment").optional().action((_, c) => c.copy(useAttachment = true))
        .text("Causes the message to be posted as an attachment in Slack")
      opt[String]("title").valueName("TITLE").optional().action((x, c) => c.copy(title = Option(x)))
        .text("Sets the attachment title")
      opt[String]("color").valueName("GOOD or WARNING or DANGER").optional().action((x, c) => c.copy(color = Option(x)))
        .text("Sends the attachment in chosen color")
      arg[String]("MESSAGE").required().unbounded().action((x, c) => c.copy(message = c.message :+ x))
        .text("Sends the given MESSAGE")
    }

    if (!parser.parse(args, Options()).exists(apply)) {
      System.exit(3)
    }
  }
}
```

**Tested for Existing Behavior**

```scala 
def main(arg: Array[String]): Unit = {
    var args = new Array[String](3)
    args(0) = "--webhook-url"
    args(1) = "my web hook url"
    args(2) = "I KNOW IT FAILED!"

    val parser = new OptionParser[Options](s"hyperion-notification-slack-activity") {
      override def showUsageOnError = true

      note("Sends a notification message to a Slack incoming webhook.")
      help("help").text("prints this usage text")
      opt[Unit]("fail-on-error").optional().action((_, c) => c.copy(failOnError = true))
        .text("Causes the activity to fail if any error received from the webhook")
      opt[String]("webhook-url").valueName("WEBHOOK").required().action((x, c) => c.copy(webhookUrl = x))
        .text("Sends the message to the given WEBHOOK url")
      opt[String]("user").valueName("NAME").optional().action((x, c) => c.copy(user = Option(x)))
        .text("Sends the message as the user with NAME")
      opt[String]("emoji").valueName("EMOJI").optional().action((x, c) => c.copy(iconEmoji = Option(x)))
        .text("Use EMOJI for the icon")
      opt[String]("to").valueName("CHANNEL or USERNAME").optional().action((x, c) => c.copy(channel = Option(x)))
        .text("Sends the message to #CHANNEL or @USERNAME")
      opt[Unit]("useAttachment").optional().action((_, c) => c.copy(useAttachment = true))
        .text("Causes the message to be posted as an attachment in Slack")
      opt[String]("title").valueName("TITLE").optional().action((x, c) => c.copy(title = Option(x)))
        .text("Sets the attachment title")
      opt[String]("color").valueName("GOOD or WARNING or DANGER").optional().action((x, c) => c.copy(color = Option(x)))
        .text("Sends the attachment in chosen color")
      arg[String]("MESSAGE").required().unbounded().action((x, c) => c.copy(message = c.message :+ x))
        .text("Sends the given MESSAGE")
    }

    if (!parser.parse(args, Options()).exists(apply)) {
      System.exit(3)
    }
  }
}
```

### CODE REVIEW

File: contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala [hunk 1]
```diff
@@ -1,20 +1,40 @@
...
+    text: String,
+    fallback: String,
+    color: String
+  )
+
+  case class SlackMessage(
+     user: Option[String] = None,
+     iconEmoji: Option[String] = None,
+     channel: Option[String] = None,
+     text: String = "",
```
Comment (ssharma) [inline contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala]: I didn't want to change legacy behavior. Can change if the author/ Kexin requires

File: contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala [hunk 2]
```diff
@@ -27,21 +47,18 @@ object SendSlackMessage {
...
-      val message = Seq(
-        "icon_emoji" -> options.iconEmoji,
-        "channel" -> options.channel,
-        "username" -> options.user,
-        "text" -> Option(options.message.mkString("\n"))
-      ).flatMap {
-        case (k, None) => None
-        case (k, Some(v)) => Option(k -> JString(v))
+      // Build new line separated user defined list of messages
+      val text = options.message.mkString("\n")
```
Comment (ssharma) [inline contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala]: Legacy behavior. Maybe, for attachment we can use multi messages. Let me do that

**Comment by jnewman** [comment]
I actually like Scalariform pretty well, but I'm not an owner of this project, so I'd leave whether that's worth it to Kexin.

File: contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala [hunk 3]
```diff
@@ -1,20 +1,40 @@
 package com.krux.hyperion.contrib.activity.notification
 
-import java.net.{ HttpURLConnection, URL }
+import java.net.{HttpURLConnection, URL}
```
Comment (realstraw) [inline contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala]: We used to. But we've agreed to change to no-padding (IDE default) to make it easier for other people to contribute.

File: contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala [hunk 4]
```diff
@@ -66,6 +108,12 @@ object SendSlackMessage {
         .text("Use EMOJI for the icon")
       opt[String]("to").valueName("CHANNEL or USERNAME").optional().action((x, c) => c.copy(channel = Option(x)))
         .text("Sends the message to #CHANNEL or @USERNAME")
+      opt[Unit]("useAttachment").optional().action((_, c) => c.copy(useAttachment = true))
+        .text("Causes the message to be posted as an attachment in Slack")
+      opt[String]("title").valueName("TITLE").optional().action((x, c) => c.copy(title = Option(x)))
```
Comment (realstraw) [inline contrib/activity/notification/src/main/scala/com/krux/hyperion/contrib/activity/notification/SendSlackMessage.scala]: does it mean if `--useAttachment` is not specified, title and color will be useless/ignored?

**Comment by realstraw** [comment]
Regarding to coding style, we don't have one that's open to others, but the internal coding style still applies. We should probably open our Scala coding style sometime.

**Comment by ssharma** [comment]
@realstraw Yes! I'll modify the client now. Maintained backward compatibility so that existing clients won't break. Let me add it to this PR itself

**Comment by realstraw** [comment]
@ssharma I'm closing this for inactivity.

**Pull request closed without merging**